File: Iteration-2/src/Advisor.java
import java.util.ArrayList;
import org.apache.log4j.Logger;

public class Advisor extends Instructor {
    static Logger log = Logger.getLogger(Advisor.class);
    private ArrayList<Student> students;

    public Advisor(String firstName, String lastName) {
        super(firstName, lastName);
        this.students = new ArrayList<Student>();
        log.info(this.getFirstName() + " " + this.getLastName() + " named advisor created.");
    }

    public Advisor(String firstName, String lastName, ArrayList<Course> givenCourses) {
        super(firstName, lastName, givenCourses);
        this.students = new ArrayList<Student>();
        log.info(this.getFirstName() + " " + this.getLastName() + " named advisor created.");
    }

    public void completeRegistration(Student student) {
        for (int i = 0; i < student.getRequestedCourses().size(); i++)
            if (checkQuotaForRegistration(student.getRequestedCourses().get(i), student)
                    && checkCollision(student, student.getRequestedCourses().get(i))) {
                student.getTranscript().getEnrolledCourses().add(student.getRequestedCourses().get(i));
                student.getRequestedCourses().get(i).getStudents().add(student);
                /* log.info(student.getStudentName() + " enrolled this course: "
                        + student.getRequestedCourses().get(i).getCourseName()); */
            }
        log.info(student.getStudentName() + "'s registration successfully completed.");
    }

    public boolean checkCollision(Student student, Course course) {
        boolean isTrue = true;
        for (Course c : student.getTranscript().getEnrolledCourses()) {
            if (course.getCourseSchedule().isCollision(c.getCourseSchedule())) {
                course.setCollisionProblem(course.getCollisionProblem() + 1);
                student.getStudentOutput().add("Advisor didn't approve " + course.getCourseCode() +
                        " because of two hours collision with " + c.getCourseCode() + " in schedule");
                log.info(student.getStudentName() + " couldn't take " + course.getCourseName()
                        + " because of collision with " + c.getCourseName() + ".");
                isTrue = false;
                break;
            }
        }
        /* log.info("No Collision found. " + student.getStudentName() + " can take " + course.getCourseName() + "."); */
        return isTrue;
    }

    public boolean checkQuotaForRegistration(Course course, Student student) {
        if (course.getStudents().size() < course.getQuato()) {
            /* log.info(
                    "Course quota not full. " + student.getStudentName() + " can take " + course.getCourseName() + "."); */
            return true;
        } else {
            course.setQuotaProblem(course.getQuotaProblem() + 1);
            student.getStudentOutput().add("The student couldn't register for " + course.getCourseCode() +
                    " because of a quota problem");
            log.info(student.getStudentName() + " couldn't take " + course.getCourseName()
                    + " because course quota is full ( Quota: " + course.getQuato() + ").");
            return false;
        }
    }

    // GETTER & SETTER
    public ArrayList<Student> getStudents() {
        return this.students;
    }

    public void setStudents(ArrayList<Student> students) {
        this.students = students;
        /* log.info(this.getAdvisorName() + ": advisor's student list changed."); */
    }

    public void addStudent(Student student) {
        this.students.add(student);
        /* log.info("Student " + student.getStudentName() + " added to advisor " + this.getAdvisorName() + "'s list."); */
    }

    public void removeStudent(Student student) {
        this.students.remove(student);
        /* log.info("Student " + student.getStudentName() + " removed from advisor " + this.getAdvisorName() + "'s list."); */
    }

    public String getAdvisorName() {
        return getFirstName() + " " + getLastName();
    }
}

File: Iteration-3/Code/Advisor.py
from Internship import Internship
from Instructor import Instructor
from Course import Course
from typing import List
import logging


class Advisor(Instructor):

    def __init__(self, firstName: str, lastName: str, givenCourses: List[Course] = []):
        super().__init__(firstName, lastName, givenCourses)
        self.__students = []
        logging.info(self.getFirstName() + " " + self.getLastName() +
                     " named advisor created.")

    def completeRegistration(self, student):
        for i in range(len(student.getRequestedCourses())):
            if not isinstance(student.getRequestedCourses()[i], Internship):
                if self.checkQuotaForRegistration(student.getRequestedCourses()[i], student) and self.checkCollision(student, student.getRequestedCourses()[i]) and self.checkEnrolledCourseLimit(student.getRequestedCourses()[i], student):
                    student.getTranscript().getEnrolledCourses().append(
                        student.getRequestedCourses()[i])
                    student.getRequestedCourses(
                    )[i].getStudents().append(student)

            if isinstance(student.getRequestedCourses()[i], Internship):
                student.getTranscript().getEnrolledCourses().append(
                    student.getRequestedCourses()[i])
                student.getRequestedCourses()[i].getStudents().append(student)

        logging.info(student.getStudentName() +
                     "'s registration successfully completed.")

    def checkCollision(self, student, course):
        if isinstance(course, Internship):
            return True
        isTrue = True
        for c in student.getTranscript().getEnrolledCourses():
            if course.getCourseSchedule().isCollision(c.getCourseSchedule()):
                course.setCollisionProblem(course.getCollisionProblem() + 1)
                student.getStudentOutput().append("Advisor didn't approve " + course.getCourseCode() +
                                                  " because of two hours collision with " + c.getCourseCode() + " in schedule")
                logging.info(student.getStudentName() + " couldn't take " + course.getCourseName() +
                             " because of collision with " + c.getCourseName() + ".")
                isTrue = False
                break
        return isTrue

    def checkQuotaForRegistration(self, course, student):
        if len(course.getStudents()) < course.getQuato():
            return True
        else:
            course.setQuotaProblem(course.getQuotaProblem() + 1)
            student.getStudentOutput().append("The student couldn't register for " +
                                              course.getCourseCode() + " because of a quota problem")
            logging.info(student.getStudentName() + " couldn't take " + course.getCourseName() +
                         " because course quota is full ( Quota: " + str(course.getQuato()) + ").")
            return False

    def checkEnrolledCourseLimit(self, course, student):
        if len(student.getTranscript().getEnrolledCourses()) <= 10:
            return True
        else:
            course.setEnrolledCourseLimit(course.getEnrolledCourseLimit() + 1)
            student.getStudentOutput().append("The student couldn't register for " +
                                              course.getCourseCode() + " due to exceeding the maximum number of courses that can be registered")
            logging.info(student.getStudentName() + " couldn't take " + course.getCourseName() +
                         " due to exceeding the maximum number of courses that can be registered.")
            return False

    def getStudents(self):
        return self.__students

    def setStudents(self, students):
        self.__students = students

    def addStudent(self, student):
        self.__students.append(student)

    def removeStudent(self, student):
        self.__students.remove(student)

    def getAdvisorName(self):
        return self.getFirstName() + " " + self.getLastName()
